addons/html_editor/static/src/core/clipboard_plugin.js

```diff
@@ -181,38 +195,512 @@ export class ClipboardPlugin extends Plugin {
+                        let alertnateValue = false;
+                        return match.replace(/ /g, () => {
+                            alertnateValue = !alertnateValue;
+                            const replaceContent = alertnateValue ? "\u00A0" : " ";
+                            return replaceContent;
+                        });
+                    });
+                    this.shared.domInsert(modifiedTextFragment);
+                    if (textIndex < textFragments.length) {
+                        // Break line by inserting new paragraph and
+                        // remove current paragraph's bottom margin.
+                        console.warn("Not implemented");
+                        // const p = closestElement(selection.anchorNode, "p");
+                        // if (isUnbreakable(closestBlock(selection.anchorNode))) {
+                        //     this._applyCommand("oShiftEnter");
+                        // } else {
+                        //     this._applyCommand("oEnter");
+                        //     p && (p.style.marginBottom = "0px");
+                        // }
+                    }
+                    textIndex++;
+                }
+            }
+        }
+    }
+
+    /**
+     * Prepare clipboard data (text/html) for safe pasting into the editor.
+     *
+     * @private
+     * @param {string} clipboardData
+     * @returns {Element}
+     */
+    prepareClipboardData(clipboardData) {
+        const container = document.createElement("fake-container");
+        container.append(parseHTML(this.document, clipboardData));
 
-    pasteHtml(html) {
-        const fragment = parseHTML(this.document, html);
-        this.shared.domInsert(fragment);
+        for (const tableElement of container.querySelectorAll("table")) {
+            tableElement.classList.add("table", "table-bordered", "o_table");
+        }
+
+        const progId = container.querySelector('meta[name="ProgId"]');
+        if (progId && progId.content === "Excel.Sheet") {
+            // Microsoft Excel keeps table style in a <style> tag with custom
+            // classes. The following lines parse that style and apply it to the
+            // style attribute of <td> tags with matching classes.
+            const xlStylesheet = container.querySelector("style");
+            const xlNodes = container.querySelectorAll("[class*=xl],[class*=font]");
+            for (const xlNode of xlNodes) {
+                for (const xlClass of xlNode.classList) {
+                    // Regex captures a CSS rule definition for that xlClass.
+                    const xlStyle = xlStylesheet.textContent
+                        .match(`.${xlClass}[^{]*{(?<xlStyle>[^}]*)}`)
+                        .groups.xlStyle.replace("background:", "background-color:");
+                    xlNode.setAttribute("style", xlNode.style.cssText + ";" + xlStyle);
+                }
+            }
+        }
```

Comment:
Maybe Excel specific code should live somewhere else?

Reply:
I added a comment